tests/utils/ExecutorTestBuilder.cpp

```diff
@@ -0,0 +1,260 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ExecutorTestBuilder.h"
+#include "ExecutorBenchmarkWrapper.h"
+#include "ExecutorUnitTestWrapper.h"
+#include "TestDeviceManager.h"
+
+using namespace glow;
+using namespace glow::runtime;
+
+void ExecutorTestBuilder::addNode(const std::string &name, DeviceIDTy deviceId,
+                                  llvm::ArrayRef<llvm::StringRef> parents,
+                                  llvm::ArrayRef<llvm::StringRef> inputs,
+                                  llvm::ArrayRef<llvm::StringRef> outputs,
+                                  RunIdentifierTy runId, bool success) {
+  auto newNode = llvm::make_unique<DAGNode>();
+  auto *newNodeRawPtr = newNode.get();
+
+  // If this is the first node being added, record the run ID for the graph.
+  // Otherwise, make sure that the runId matches that of the previous nodes.
+  if (nodes_.empty()) {
+    runId_ = runId;
+  } else {
+    assert(runId == runId_ && "Node run ID does not match rest of graph!");
+  }
+
+  // If the result for this node is false, set the expected
+  // result for the entire test to false.
+  success_ &= success;
+
+  // Add parents to the list of parents in the new node and add the newNode
+  // to the list of children in the parents. If the parent list is empty,
+  // make the root the only parent. Also, update the set of known leaves
+  // by removing any parents of the new node from it. This will be useful
+  // later.
+  if (!parents.empty()) {
+    for (const auto &parent : parents) {
+      auto it = nodes_.find(parent);
+      if (it == nodes_.end()) {
+        assert(!"Parent specified for node not found!");
+      }
+      DAGNode *parentPtr = (it->second).get();
+      (newNode->parents).emplace_back(parentPtr);
+      (parentPtr->children).emplace_back(newNodeRawPtr);
+      leaves_.erase(parentPtr);
+    }
+  } else {
+    (newNode->parents).emplace_back(root_.get());
+    (root_->children).emplace_back(newNode.get());
+  }
+
+  // Iterate through inputs and outputs and:
```

Comment:
Is this normally done by the partitioner/provisioner? Can we reuse that code?

Reply:
@gcatron Can we?

Reply:
I think we're short circuiting that logic here. The partitioner/provisioner reason over a network graph. This is creating the DAG directly.